source/vstd/bits.rs

```diff
@@ -204,3 +208,112 @@ lemma_shl_is_mul!(lemma_u64_shl_is_mul, lemma_u64_shl_is_mul_auto, lemma_u64_pow
 lemma_shl_is_mul!(lemma_u16_shl_is_mul, lemma_u16_shl_is_mul_auto, lemma_u16_pow2_no_overflow, u16);
 lemma_shl_is_mul!(lemma_u8_shl_is_mul, lemma_u8_shl_is_mul_auto, lemma_u8_pow2_no_overflow, u8);
+
+// Proofs that and with mask is equivalent to modulo with power of two.
+macro_rules! lemma_mask_is_mod {
+    ($name:ident, $name_auto:ident, $and_split_low_bit:ident, $no_overflow:ident, $uN:ty) => {
+        #[cfg(verus_keep_ghost)]
+        verus! {
+        #[doc = "Proof that for given natural n and x of type "]
+        #[doc = stringify!($uN)]
+        #[doc = ", and with the low n-bit mask is equivalent to modulo 2^n."]
+        pub proof fn $name(x: $uN, n: nat)
+            requires
+                n < <$uN>::BITS,
+            ensures
+                x & (mask(n) as $uN) == x % (pow2(n) as $uN),
+            decreases n,
+        {
+            // Bounds.
+            $no_overflow(n);
+            lemma_pow2_pos(n);
+
+            // Inductive proof.
+            if n == 0 {
+                assert(mask(0) == 0) by (compute_only);
+                assert(x & 0 == 0) by (bit_vector);
+                assert(pow2(0) == 1) by (compute_only);
+                assert(x % 1 == 0);
+            } else {
+                lemma_pow2_unfold(n);
+                assert((x % 2) == ((x % 2) & 1)) by (bit_vector);
+                calc!{ (==)
+                    x % (pow2(n) as $uN);
+                        {}
+                    x % ((2 * pow2((n-1) as nat)) as $uN);
+                        {
+                            lemma_pow2_pos((n-1) as nat);
+                            lemma_mod_breakdown(x as int, 2, pow2((n-1) as nat) as int);
+                        }
+                    add(mul(2, (x / 2) % (pow2((n-1) as nat) as $uN)), x % 2);
+                        {
+                            $name(x/2, (n-1) as nat);
+                        }
+                    add(mul(2, (x / 2) & (mask((n-1) as nat) as $uN)), x % 2);
+                        {
+                            lemma_mask_div2(n);
+                        }
+                    add(mul(2, (x / 2) & (mask(n) as $uN / 2)), x % 2);
+                        {
+                            lemma_mask_is_odd(n);
+                        }
+                    add(mul(2, (x / 2) & (mask(n) as $uN / 2)), (x % 2) & ((mask(n) as $uN) % 2));
+                        {
+                            $and_split_low_bit(x as $uN, mask(n) as $uN);
+                        }
+                    x & (mask(n) as $uN)
+                }
+            }
+        }
+
+        #[doc = "Proof that for all natural n and x of type "]
+        #[doc = stringify!($uN)]
+        #[doc = ", and with the low n-bit mask is equivalent to modulo 2^n."]
+        pub proof fn $name_auto()
+            ensures
+                forall|x: $uN, n: nat| n < <$uN>::BITS ==> #[trigger] (x & (mask(n) as $uN)) == (x % (pow2(n) as $uN)),
```

Comment:
I was unsure how to set triggers for this. At first I tried both sides of the equality, but that was disallowed due to an error about not being able to configure triggers for both arithmetic and bitwise operations on `x`.

Reply:
I think triggering on the LHS is preferable.  It seems unlikely that you'll care about this relation just because you perform a mod operation w.r.t. pow2, but it seems quite likely that you'll care if you're doing the bitwise operation.

Reply:
Okay, sounds good thanks.